src/helpers.ts

```diff
@@ -1,24 +1,120 @@
+  PartialPageObjectResponse,
+  PartialUserObjectResponse,
+  UserObjectResponse,
+} from "./api-endpoints"
+
+interface PaginatedArgs {
+  start_cursor?: string
+}
+
+interface PaginatedList<T> {
+  object: "list"
+  results: T[]
+  next_cursor: string | null
+  has_more: boolean
+}
+
+/**
+ * Returns an async iterator over the results of any paginated Notion API.
+ *
+ * Example (given a notion Client called `notion`):
+ *
+ * ```
+ * for await (const block of iteratePaginatedAPI(notion.blocks.children.list, {
+ *   block_id: parentBlockId,
+ * })) {
+ *   // Do something with block.
+ * }
+ * ```
+ *
+ * @param listFn A bound function on the Notion client that represents a conforming paginated
+ *   API. Example: `notion.blocks.children.list`.
+ * @param firstPageArgs Arguments that should be passed to the API on the first and subsequent
+ *   calls to the API. Any necessary `next_cursor` will be automatically populated by
+ *   this function. Example: `{ block_id: "<my block id>" }`
+ */
+export async function* iteratePaginatedAPI<Args extends PaginatedArgs, Item>(
+  listFn: (args: Args) => Promise<PaginatedList<Item>>,
+  firstPageArgs: Args
+): AsyncIterableIterator<Item> {
+  let nextCursor: string | null | undefined = firstPageArgs.start_cursor
+  do {
+    const response: PaginatedList<Item> = await listFn({
+      ...firstPageArgs,
+      start_cursor: nextCursor,
+    })
+    yield* response.results
+    nextCursor = response.next_cursor
+  } while (nextCursor)
+}
+
 /**
- * Utility for enforcing exhaustiveness checks in the type system.
+ * Collect all of the results of paginating an API into an in-memory array.
  *
- * @see https://basarat.gitbook.io/typescript/type-system/discriminated-unions#throw-in-exhaustive-checks
+ * Example (given a notion Client called `notion`):
  *
- * @param value The variable with no remaining values
+ * ```
+ * const blocks = collectPaginatedAPI(notion.blocks.children.list, {
+ *   block_id: parentBlockId,
+ * })
+ * // Do something with blocks.
+ * ```
+ *
+ * @param listFn A bound function on the Notion client that represents a conforming paginated
+ *   API. Example: `notion.blocks.children.list`.
+ * @param firstPageArgs Arguments that should be passed to the API on the first and subsequent
+ *   calls to the API. Any necessary `next_cursor` will be automatically populated by
+ *   this function. Example: `{ block_id: "<my block id>" }`
+ */
+export async function collectPaginatedAPI<Args extends PaginatedArgs, Item>(
+  listFn: (args: Args) => Promise<PaginatedList<Item>>,
+  firstPageArgs: Args
+): Promise<Item[]> {
+  const results: Item[] = []
+  for await (const item of iteratePaginatedAPI(listFn, firstPageArgs)) {
+    results.push(item)
+  }
+  return results
+}
+
+/**
+ * @returns `true` if `response` is a full `BlockObjectResponse`.
  */
-export function assertNever(value: never): never {
-  throw new Error(`Unexpected value should never occur: ${value}`)
+export function isFullBlock(
```

Comment:
i considered making these more type safe with e.g. `makeTypeGuard`, but for some complicated typescript reasons it seems that this didn't actually buy us any type safety. check out [this typescript playground](https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBAYwgOwM7wLYEMDWwAqAnmMAOICuWUAJnALwBQAkADzNMDqAltQccABp2AZXIIEwVKiIk4oGMGTVUcbrxmD2AMSxcANhrkgFSlWr6y6cAKIgEe8rxbmNAuKPGTp-AHzMfABTsMPwUVNQAXHABXMhg5DBRLvwAlPQ+cADecDBQ5MBRHhJShgC+cAA+WXAAZlh6qAVwOvplzGl0fkwBAG71+Uk8FsApUX0OwHBcKkVehp3sAES5+YtTyDmhlDS9-SMMByGyWuR6egAKWADmk1bZPFHoULFXANxw5FB6j7kvcKUMI6TS6wLj1S43ejVB5wJ5-AEMJBoeDTE5nCG3ODYPAaMI0Fhoi7XSZVEEwMFEm5uQkYwLMCAAIwAVuloosGRAruRUGtYnBGSyAPSCuAAIQA8uKALJRGAAC2mcHqek2sgQcuACBwqDcGGAWGQfwA7pMbvBkBAldRqFxySh6qrJqgsDVgCFalAIBgsbhhnjqAw4EHg0GAPzVFZNAX-ODC-nMqYqIG0LAqRbIYA9YBQNYaqCTI0K9WJpVgMBURQwPSEJVSLhXZBYBl6SYwS0Gwjyv4xZD2Rx-SMpQMh4NRbJ1BpRhOlIcpIA) for the example -- an incorrect makeTypeGuard that still compiles.

i think if we changed PartialBlockObject to have a `type?: undefined`, we could change the guard to just be on the truthiness of `response.type` -- but i don't really want to update codegen just for this 😔